### ISSUE TITLE
Fibaro Dimmer v2 removing - "unhandled error" + UNREMOVABLE ghost device in list and at devices [0.8.17]

### ISSUE THREAD

**Issue by athomey** [athombv/homey#167]
When removing it (I only pressed the button ONCE during this boot cycle of Homey), it starts complaining 
![image](https://cloud.githubusercontent.com/assets/16608801/13021090/6c9e9efa-d1d9-11e5-9526-8d8c63200684.png)

I also have a ghost Device now at the devices tab:

![image](https://cloud.githubusercontent.com/assets/16608801/13021156/c90aa968-d1d9-11e5-9fd8-0aad8387bef7.png)

And also in the dropdown at zwave settings a ghost-device:

![image](https://cloud.githubusercontent.com/assets/16608801/13021172/e53a9b3e-d1d9-11e5-871f-e0deb71d82de.png)

#168 was _before_ this problem, so maybe there is a connection.

When trying to remove the failed node (I already factory reset the Dimmer itself, cause removing normally wouldn't work after above errors anymore in Homey), I get the following screen for minutes:

![image](https://cloud.githubusercontent.com/assets/16608801/13021374/299720da-d1db-11e5-8d14-42d306211639.png)

After a while I get:

![image](https://cloud.githubusercontent.com/assets/16608801/13021387/390d210e-d1db-11e5-9c79-d3e83e9f99e8.png)

The state is now:

![image](https://cloud.githubusercontent.com/assets/16608801/13021410/572cc82e-d1db-11e5-9a8c-00ceb73a5736.png)

I will reboot and retry removing the ghost nodes.

((Sidenote: This smells like an atomic transaction code smell, and/or racing condition or threading problem...))

**Comment by athomey** [comment]
After rebooting, still ghost device (and message gets outside the window:

![image](https://cloud.githubusercontent.com/assets/16608801/13021510/f0310d5a-d1db-11e5-9305-5f5d3104748b.png)

also still in device dropdown for failed nodes:

![image](https://cloud.githubusercontent.com/assets/16608801/13021526/06f1228c-d1dc-11e5-8ee1-6b26c666c2d5.png)

I tried Remove Failed Node, and see this for minutes now:

![image](https://cloud.githubusercontent.com/assets/16608801/13021539/22061302-d1dc-11e5-8609-a2280b779ace.png)

((Remark about above screenshot, why is the dropdown and Refresh and REplace button not disabled? I can trigger even more code to run, while the Remove node button still shows it's doing something. This is asking for multiple-entry or multi-threading errors and racing conditions. 2 cts.))

After aforementioned waiting for minutes, the UI shows first this error again:

![image](https://cloud.githubusercontent.com/assets/16608801/13021736/19dfdd92-d1dd-11e5-9a5a-5b1c2e90c354.png)

and then returns to normal, no more snake-eat-tail hourglasses, but my ghost-node, nor my ghost device have been removed.... I am out of options....

The ball is now in your court @ocatie   :-)  Thx for your effort!

**Comment by ocatie** [comment]
That is an unreachable device. A ghost device/ghost node is a registered node id in the network and the corresponding device is gone. This is mostly the case on an error "TRANSMIT_COMPLETE_NO_ACK". With an "timeout waiting for report" the data is acked by the device but the command isn't answered. Try selecting the device in the dropdown in the settings page and press the "Refresh node" button. If the refresh is successful the device will become available again. If not, you should be able to remove the device using the "Remove Failed Node button" (since we just verified the device is unreachable).

If you receive an "An error occured true" the connection with the driver has been lost. Probably because of a crash. You'll have to reboot Homey and try again. Please try to reproduce this message after a reboot.

If you only get the "Running in another session" message you can try excluding your device (to finish the session) or press the "Restart Z-Wave chip" button. This will restart the chip and clear all running sessions (except on a crash because... a crash).

**Comment by athomey** [comment]
Thx for your swift response again! 

It _is_ a ghost device persé, cause the device couldn't be removed (due to have added nodes I guess), after that I factory reset the device, so it doesn't exist anymore "in the air". It will be unreachable persé, after that "refresh Node" has no use, cause it's not associated anymore. I think (I hope I) gave you perfect reproduction of this problem of never-to-be-able-to-remove-nodes and ghost-device. I already rebooted the whole of Homey, before trying again, so I suspect that also has reset the zwave chip. ;-)

((It looks a bit like https://github.com/athombv/homey/issues/14 and https://github.com/athombv/homey/issues/19 ))

**Comment by athomey** [comment]
Here you suggest that issue 167 is a crashing driver, am I understanding you correctly? >> https://github.com/athombv/homey/issues/129#issuecomment-183515386  Thx for not closing this issue btw, you are a bit fast in closing them. ;-)

My idea is (cause I saw it a few times now, and only got it fixed by doing factory reset + data clear of homey itself) that the node- errors on adding, damage something in your DB/data-consistency, and then preventing me from deleting the nodes/devices, thus rendering them as ghost-items. ((The atomic transaction discussion we had.)) What are your thoughts on that angle @ocatie ?

**Comment by ocatie** [comment]
My thoughts at the moment are that the remove node problems are mostly caused by the front end. The internal Z-Wave API is async and only one process can run at the time. This is something out of my control. At the moment the frontend isn't synced correctly with the current state of the driver so processes can be started in the frontend but aren't in the driver and the returned error messages are too generic.

Also I think there is still might be a crash somewhere so with every button click processes are started without being started and more de-syncing occurs. But the problem is I'm not able to produce the crash I think exists (which is good in a way because more chance there is no crash after all).

Long story short. It's very hard to find the cause of the issues without an proper synced interface (I'm not a frontender ;) ) so first I'm implementing a event system for the settings page.
I'm planning to redo a lot of the settings page but it will take some time. Also I still have to look out for the possible crash so that I'm not just conquering symptoms.

For now be wary not to different buttons too quick. When the Remove node process is running the Restart Z-Wave chip process can be run to end all processes. If the process still runs after a refresh the settings screen you have to reboot.

**Comment by ocatie** [comment]
Please report back after update 0.8.18.

**Comment by ocatie** [comment]
The issues reported in this issue should be fixed in 0.8.19. I found a crash caused when a device is being deleted while interviewing causing these symptoms.

You can remove the ghost node (if the device is removed from the network already) by selecting it in the dropdown and using the Remove Failed Node button. If you get an error the id is not on the failed node list, use the Refresh Node button first..
